### PR TITLE
BIP141: New commitment structure

### PR DESCRIPTION
**Pull request by jl2012**
@sipa 

Draft based on http://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-February/012313.html

### CODE REVIEW

File: bip-0141.mediawiki [hunk 1]
```diff
@@ -58,22 +58,15 @@ A non-witness program (defined hereinafter) txin MUST be associated with an empt
...
-   4-byte - Commitment header (0xaa21a9ed)
-  32-byte - Commitment hash: Double-SHA256(witness root hash|witness nonce)
-  
-  39th byte onwards: Optional data with no consensus meaning
-  
-and the coinbase's input's witness must consist of a single 32-byte array for the witness nonce.
+For each stack item in the coinbase witness, a double SHA256 is calculated. The "witness commitment" is the root of a Merkle tree, whose leaves are all these hashes.
 
-If there are more than one scriptPubKey matching the pattern, the one with highest output index is assumed to be the commitment.
+Item 0 of the coinbase witness stack is the "wtxid commitment". The "wtxid commitment" must be exactly 32 bytes. It is the root of a Merkle tree, whose leaves are all those <code>wtxid</code> in the block, calculated in a way similar to the hashMerkleRoot in the block header. The <code>wtxid</code> of coinbase transaction is assumed to be <code>0x0000....0000</code>.
```
Comment (NicolasDorier) [inline bip-0141.mediawiki]: Item 0 mean top of the stack or bottom of the stack ? I assume bottom, but maybe it can be clarified.

Comment (jl2012) [inline bip-0141.mediawiki]: Yes, 0 means bottom in the source code

**Comment by luke-jr** [comment]
What's the status of this? Ready for merging?

**Comment by sipa** [comment]
It's an idea, and I'm waiting for more opinions first.

File: bip-0141.mediawiki [hunk 2]
```diff
@@ -178,16 +171,19 @@ Comparing with the previous example, the scriptPubKey is 11 bytes smaller (with
...
-coinbase becomes:
+The new commitment in coinbase transaction is a Merkle root of coinbase witness stack items. Currently, only the first item has consensus meaning. The rest of stack items can be used for future commitments, such as fraud proof or UTXO commitment.
+
+Any contents or commitments that are not consensus-critical to Bitcoin, such as merge-mining, MUST NOT use the coinbase witness stack to preserve the ability to do upgrades of the Bitcoin consensus protocol.
+
+=== Location of the witness commitment ===
 
-  Double-SHA256(Witness root hash|Hash(new commitment|witness nonce))
+Theorectically, the witness commitment could be placed in any location within the coinbase transaction. In practice, most miners should place it in either the scriptSig or as one of the output scriptPubKey. The following discussion assumes that the BIP34 block size push occupies 4 bytes, which is true for the following 300 years with 10 minutes block interval.
```
Comment (NicolasDorier) [inline bip-0141.mediawiki]: typo "Theorectically"

**Pull request closed without merging**